Support added for ruby 1.9.3

nitesh-varma opened the pull request:
Support added for ruby 1.9.3
Successfully executed specs both in 1.9.3 & 2.2.0

=> ruby-2.2.0 [ i686 ]
Finished in 1 minute 54.09 seconds (files took 0.31953 seconds to load)
122 examples, 0 failures, 4 pending

=* ruby-1.9.3-p551 [ i686 ]
Finished in 2 minutes 23.5 seconds (files took 0.37978 seconds to load)
122 examples, 0 failures, 4 pending

rilian commented:
so all new startups that use Plaid would suffer from old and insecure stuff in ruby version that is outdated for more than 3 years?

nitesh-varma commented:
Hello Rilian,

No, startups will definitely use ruby > 2.
Stuffs added for above will work from ruby >= 1.9.3.
Is changes done by me are insecure?

I had just updated the way Net::Http  used to sent get & post  requests such that it would work on both 1.9.3 and new versions i.e > 2.0.0

There would be many big projects that are running large application on rails from last 3 - 4 years.
These projects would be on ruby < 2.0.0, but due to compatibility issues that would not be able to use such nice and updated gem of Plaid API.

Please advise, as new startups will use latest ruby version so they would not encounter any issue of 1.9.3, but these gem can also be used by old projects which are not yet migrated in new version of ruby (due to several dependency and current value of applications in market) 

Thanks,
Nitesh

rilian commented on lib/plaid/connection.rb:
```diff
@@ -9,22 +9,29 @@ class << self
         options.merge!(client_id: Plaid.customer_id, secret: Plaid.secret)
-        res = Net::HTTP.post_form(uri,options)
+        http = Net::HTTP.new(uri.host, uri.port)
+        http.use_ssl = (uri.scheme == 'https')
```
i think Plaid always uses SSL, so this `http.use_ssl` should be just strictly `true`

nitesh-varma replied:
Hello Rilian,

Yes, but in Plaid::Connection.patch, Plaid::Connection.delete, master current code uses :use_ssl => uri.scheme == 'https' check.

>         Net::HTTP.start(uri.hostname, uri.port, :use_ssl => uri.scheme == 'https') { |http| http.request(req) }
> res = Net::HTTP.start(uri.hostname, uri.port, :use_ssl => uri.scheme == 'https') { |http| http.request(req) }

So I also added ssl check in post & get method while doing modification and not used  strictly true.

Thanks,
Nitesh

rilian replied:
Hi Nitesh , yep, I saw this. Generally your PR is fine, but I doubt anyone wants to send their bank transaction tokens via unencrypted connection, that's why I believe SSL should be `true` by default.

PS: By the way, I'm not Plaid developer, we just use Plaid in our startup...

darthrevan replied:
A quote from Plaid doc: "The Plaid API is served over HTTPS to ensure data privacy; HTTP is not supported. "
I think this http vs https logic could be safely replaced for strict http.

nitesh-varma replied:
Yes, as per https://github.com/plaid/plaid-ruby#configuring-plaid

<pre>
Plaid.config do |p|
  p.customer_id = 'Plaid provided customer ID here'
  p.secret = 'Plaid provided secret key here'
  p.environment_location = 'URL for the development or production environment'
  # i.e. 'https://tartan.plaid.com/' for development, or
  # 'https://api.plaid.com/' for production
end
</pre>

We either use 'https://tartan.plaid.com/' or 'https://api.plaid.com/'. So both the cases it would be https only and ssl will be set true.

> i think Plaid always uses SSL
> I think this http vs https logic could be safely replaced for strict http in current master code too.

Yes, we should strictly use https i.e :use_ssl => true for plaid requests.

Thanks,
Nitesh

darthrevan commented:
There are also Ruby projects that are 5-6 years old. Does it mean that Plaid API should also support Ruby 1.8?

I suggest that legacy Ruby version support should be moved to separate branch. Older project maintainers will be still able to use it by specifying a legacy support branch in their Gemfile.

nitesh-varma commented:
Hello Darthrevan,

In most of the gems I found that they supports >= ruby 1.9.3 (i.e mostly supports are removed for 1.9.2).

> https://www.ruby-lang.org/en/news/2014/11/13/ruby-1-9-3-p551-is-released/

Ruby 1.9.3's last version was released on 13 Nov 2014, which was just 8 - 9 months older, So I think we could have support of this version of ruby in master code. 

Thanks,
Nitesh

darthrevan commented:
https://www.ruby-lang.org/en/news/2015/02/23/support-for-ruby-1-9-3-has-ended/

Ruby 1.9.3 support has ended and it no longer receives important security updates.

dmerrick commented:
I think this PR seems reasonable. Sure, 1.9.3 is outdated, but if it means more people can use the project, I'm for it.

I do agree with @darthrevan that we should probably use strict HTTPS support. Plaid's backend will always be HTTPS, so there's no harm in enforcing it here.

P.S. what's wrong with girls' blog posts?

darthrevan commented:
Nothing wrong, but they are usually a little bit more public then bank account data :lipstick:

dmerrick commented:
Merged, and fixed the https issue in a separate commit.

Thanks, Nitesh!
